python/cuspatial/cuspatial/core/binpreds/feature_contains.py

```diff
@@ -154,31 +156,56 @@ def _preprocess(self, lhs, rhs):
         return _basic_equals_any(lhs, rhs)
 
 
+class MultiPointMultiPointContains(BinPred):
+    def _preprocess(self, lhs, rhs):
+        return _basic_equals_all(rhs, lhs)
+
+
 class LineStringPointContains(BinPred):
     def _preprocess(self, lhs, rhs):
         intersects = _basic_intersects(lhs, rhs)
-        equals = _basic_equals_any(lhs, rhs)
+        equals = _basic_equals_count(lhs, rhs) == rhs.sizes
         return intersects & ~equals
 
 
 class LineStringLineStringContainsPredicate(BinPred):
     def _preprocess(self, lhs, rhs):
+        # rhs_self_intersection is rhs where all colinear segments
+        # are compacted into a single segment. This is necessary
+        # because the intersection of lhs and rhs below will
+        # compact colinear segments into a single segment, so
+        # vertices in rhs will be lost and not countable.
+        # Consider the following example:
+        # lhs = [(0, 0), (1, 1)]
+        # rhs = [(0, 0), (0.5, 0.5), (1, 1)]
+        # The intersection of lhs and rhs is [(0, 0), (1, 1)].
+        # rhs is contained by lhs if all of the vertices in the
+        # intersection are in rhs. However, the intersection
+        # does not contain the vertex (0.5, 0.5) in the original rhs.
+        rhs_self_intersection = _basic_intersects_pli(rhs, rhs)
```

Comment:
It feels like you need a separate API that handles linestring merging. Something like:

```c++
void merge_points_on_multilinestring_range(MultiLinestringRange range, OutputIt output);
```

Because calling the self intersection just to remove the additional points are too expensive.